MBT Simple pendulum plant

amcastro-tri opened the pull request:
Downgrade from the n-link pendulum in #7131. To be used as a first experiment for:
- Wiring up GeometrySystem with a MultibodyTree based plant.
- Parsing a multibody/geometry system.
- Geometry registration scheme.
- Scalar conversion.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/7172)
<!-- Reviewable:end -->

amcastro-tri commented:
@seancurtis-tri for feature review

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-:-KvZEEj8LDxpfnuazlB9:bdlfnpf)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
@amcastro-tri FYI With the updates happening to `master` these days, I suggest you merge up and re-push this PR, else it will be come stale and more difficult to refresh later.

amcastro-tri commented:
We we'll be working on this PR with @SeanCurtis-TRI this week now that I am at HQ. Thanks @jwnimmer-tri.

amcastro-tri commented:
+@seancurtis-tri for feature review please.

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-:-L-8FAadX1jaKqtfFNMq:b-4269ug)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Reviewed 8 of 8 files at r2.
Review status: all files reviewed at latest revision, 27 unresolved discussions.

---

*[drake/examples/multibody/pendulum/BUILD.bazel, line 34 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-8y42MjcKge2UCcab8:-L-8y42MjcKge2UCcab9:b-lic5af) ([raw file](https://github.com/robotlocomotion/drake/blob/16110aa9a756b18ab09c9ca2240cd44113235fc4/drake/examples/multibody/pendulum/BUILD.bazel#L34)):*
> ```
>     deps = [
>         ":pendulum_plant",
>         "//drake/geometry:geometry_system",
> ```

FYI Dependency on `:pendulum_plant` already grants access to `geometry_system`.

---

*[drake/examples/multibody/pendulum/BUILD.bazel, line 37 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-9-LgcFLVroondMZGA:-L-9-LgcFLVroondMZGB:b1be386) ([raw file](https://github.com/robotlocomotion/drake/blob/16110aa9a756b18ab09c9ca2240cd44113235fc4/drake/examples/multibody/pendulum/BUILD.bazel#L37)):*
> ```
>         "//drake/geometry:geometry_visualization",
>         "//drake/systems/analysis:implicit_euler_integrator",
>         "//drake/systems/analysis:runge_kutta2_integrator",
> ```

BTW unused

---

*[drake/examples/multibody/pendulum/BUILD.bazel, line 43 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-9-TybERbSAs7OA1ga:-L-9-TybERbSAs7OA1gb:b1be386) ([raw file](https://github.com/robotlocomotion/drake/blob/16110aa9a756b18ab09c9ca2240cd44113235fc4/drake/examples/multibody/pendulum/BUILD.bazel#L43)):*
> ```
>         "//drake/systems/framework:diagram",
>         "//drake/systems/lcm",
>         "//drake/systems/primitives:constant_vector_source",
> ```

BTW unused

---

*[drake/examples/multibody/pendulum/BUILD.bazel, line 44 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-9-WY0XDXBULIWPWOb:-L-9-WY0XDXBULIWPWOc:b1be386) ([raw file](https://github.com/robotlocomotion/drake/blob/16110aa9a756b18ab09c9ca2240cd44113235fc4/drake/examples/multibody/pendulum/BUILD.bazel#L44)):*
> ```
>         "//drake/systems/lcm",
>         "//drake/systems/primitives:constant_vector_source",
>         "//drake/systems/primitives:signal_logger",
> ```

BTW unused

---

*[drake/examples/multibody/pendulum/passive_simulation.cc, line 12 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-8yWYgsCpJukrU8UGl:-L-8yWYgsCpJukrU8UGm:b-84iu82) ([raw file](https://github.com/robotlocomotion/drake/blob/16110aa9a756b18ab09c9ca2240cd44113235fc4/drake/examples/multibody/pendulum/passive_simulation.cc#L12)):*
> ```cpp
> #include "drake/lcmt_viewer_draw.hpp"
> #include "drake/systems/analysis/implicit_euler_integrator.h"
> #include "drake/systems/analysis/runge_kutta2_integrator.h"
> ```

Unused? It's not included in your command-line options.

---

*[drake/examples/multibody/pendulum/passive_simulation.cc, line 28 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-8yNpjh_BfdgMY3ZUj:-L-8yNpjh_BfdgMY3ZUk:b-4mojld) ([raw file](https://github.com/robotlocomotion/drake/blob/16110aa9a756b18ab09c9ca2240cd44113235fc4/drake/examples/multibody/pendulum/passive_simulation.cc#L28)):*
> ```cpp
> 
> DEFINE_double(target_realtime_rate, 1.0,
>               "Playback speed.  See documentation for "
> ```

FYI "playback speed" seems to be a misnomer (particularly given the flag name).

---

*[drake/examples/multibody/pendulum/passive_simulation.cc, line 48 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-8ygm6QQpDwewqP_pO:-L-8ygm6QQpDwewqP_pP:b-f3vdj8) ([raw file](https://github.com/robotlocomotion/drake/blob/16110aa9a756b18ab09c9ca2240cd44113235fc4/drake/examples/multibody/pendulum/passive_simulation.cc#L48)):*
> ```cpp
>   systems::DiagramBuilder<double> builder;
> 
>   auto geometry_system = builder.AddSystem<GeometrySystem<double>>();
> ```

BTW `auto*` would make it more apparent that `geometry_system` is a pointer.

Alternatively,

`auto& geometry_system = *builder.AddSystem<GeometrySYstem<double.>();` 

means future invocations of it being passed would look like: `&geometry_system` (consistent with our whole "return a reference" movement).

---

*[drake/examples/multibody/pendulum/passive_simulation.cc, line 128 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-8zhTIJrCW0Z56o1tH:-L-8zhTJ-1cfcM2NV48o:b-nbnkmc) ([raw file](https://github.com/robotlocomotion/drake/blob/16110aa9a756b18ab09c9ca2240cd44113235fc4/drake/examples/multibody/pendulum/passive_simulation.cc#L128)):*
> ```cpp
>   } else {
>     throw std::logic_error(
>         "Integration scheme not supported for this example.");
> ```

FYI Might be worth including the actual value of the string you got, particularly as this is case sensitive.

---

*[drake/examples/multibody/pendulum/pendulum_plant.cc, line 16 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-8mNdMlfdRHCHlQ8Fd:-L-8mNdMlfdRHCHlQ8Fe:b-2ovz7b) ([raw file](https://github.com/robotlocomotion/drake/blob/16110aa9a756b18ab09c9ca2240cd44113235fc4/drake/examples/multibody/pendulum/pendulum_plant.cc#L16)):*
> ```cpp
> #include "drake/multibody/multibody_tree/uniform_gravity_field_element.h"
> #include "drake/systems/framework/leaf_system.h"
> #include "drake/systems/rendering/pose_bundle.h"
> ```

BTW Unused?

---

*[drake/examples/multibody/pendulum/pendulum_plant.cc, line 64 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-8okUf_wIo5qwmI-rN:-L-8okUf_wIo5qwmI-rO:bggl0mc) ([raw file](https://github.com/robotlocomotion/drake/blob/16110aa9a756b18ab09c9ca2240cd44113235fc4/drake/examples/multibody/pendulum/pendulum_plant.cc#L64)):*
> ```cpp
>   DRAKE_DEMAND(geometry_system != nullptr);
>   RegisterGeometry(geometry_system);
>   DeclareGeometrySystemPorts();
> ```

Seems that this happens twice -- once here, and once in the next constructor (which is indirectly invoked by *this* constructor). 

---

*[drake/examples/multibody/pendulum/pendulum_plant.cc, line 75 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-8o6Quqi5BjvtrNl0o:-L-8o6Quqi5BjvtrNl0p:b-9ih06h) ([raw file](https://github.com/robotlocomotion/drake/blob/16110aa9a756b18ab09c9ca2240cd44113235fc4/drake/examples/multibody/pendulum/pendulum_plant.cc#L75)):*
> ```cpp
>     mass_(mass), length_(length), gravity_(gravity),
>     source_id_(source_id), frame_id_(frame_id) {
>   // Build the MultibodyTree model for this plant.
> ```

FYI This comment seems overkill; it's almost a word-for-word repeat of the function name.

---

*[drake/examples/multibody/pendulum/pendulum_plant.cc, line 87 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-8pA7pfrjtkMlolMeR:-L-8pA7pfrjtkMlolMeS:br985bi) ([raw file](https://github.com/robotlocomotion/drake/blob/16110aa9a756b18ab09c9ca2240cd44113235fc4/drake/examples/multibody/pendulum/pendulum_plant.cc#L87)):*
> ```cpp
>       model_->get_num_velocities(), 0 /* num_z */);
> 
>   DeclareGeometrySystemPorts();
> ```

As this constructor knows nothing about geometry, should the output ports exist if it isn't registering geometry?

---

*[drake/examples/multibody/pendulum/pendulum_plant.cc, line 92 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-8pQgI03Dfhc5sM6Hi:-L-8pLuHERtL6WiYydAg:b-rhgrqg) ([raw file](https://github.com/robotlocomotion/drake/blob/16110aa9a756b18ab09c9ca2240cd44113235fc4/drake/examples/multibody/pendulum/pendulum_plant.cc#L92)):*
> ```cpp
> template<typename T>
> void PendulumPlant<T>::DeclareGeometrySystemPorts() {
>   if (frame_id_.is_valid()) {
> ```

BTW Related to my previous comment; if this is only invoked when a frame is *known* to exist, this test becomes superfluous.

---

*[drake/examples/multibody/pendulum/pendulum_plant.cc, line 144 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-8q7jbuprkXDi3uKF0:-L-8psd7JrCY3HIh0HC0:b-89h6rm) ([raw file](https://github.com/robotlocomotion/drake/blob/16110aa9a756b18ab09c9ca2240cd44113235fc4/drake/examples/multibody/pendulum/pendulum_plant.cc#L144)):*
> ```cpp
> void PendulumPlant<T>::CalcFramePoseOutput(
>     const Context<T>&, FramePoseVector<T>*) const {
>   DRAKE_ABORT_MSG("There is no implementation for T != double.");
> ```

FYI Is this an older relic? `GeometrySystem` in master *should* support `AutoDiffXd`.

---

*[drake/examples/multibody/pendulum/pendulum_plant.cc, line 180 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-8qYJ_v4tz9elsWVF-:-L-8qYJ_v4tz9elsWVF0:b-28cgam) ([raw file](https://github.com/robotlocomotion/drake/blob/16110aa9a756b18ab09c9ca2240cd44113235fc4/drake/examples/multibody/pendulum/pendulum_plant.cc#L180)):*
> ```cpp
>         other.get_length(),
>         other.get_gravity(),
>         other.source_id_, other.frame_id_) {}
> ```

FYI You have three constructors. The private constructor exists so you can do this. Alternatively, you could eliminate the private constructor if you move its contents into the three-parameter constructor and then modify this so that it's body is:

```
source_id_ = other.source_id_;
frame_id_ = other.frame_id_;
```

The whole extra constructor and additional level of indirection seems a bit much to avoid these two lines.

---

*[drake/examples/multibody/pendulum/pendulum_plant.cc, line 189 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-8v9DAdzRpDPDRgWge:-L-8v9DAdzRpDPDRgWgf:b5ooa5h) ([raw file](https://github.com/robotlocomotion/drake/blob/16110aa9a756b18ab09c9ca2240cd44113235fc4/drake/examples/multibody/pendulum/pendulum_plant.cc#L189)):*
BTW The phrase: "where the rotation axis is" met unhappiness over here. How about:

> ...at the world's origin Wo, through which the rotation axis passes.

---

*[drake/examples/multibody/pendulum/pendulum_plant.h, line 18 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-8gsKqJZTKEeY-IqRI:-L-8gsKqJZTKEeY-IqRJ:b-gqd2ci) ([raw file](https://github.com/robotlocomotion/drake/blob/16110aa9a756b18ab09c9ca2240cd44113235fc4/drake/examples/multibody/pendulum/pendulum_plant.h#L18)):*
> ```objc
> 
> /// A model of an idealized pendulum with a point mass on the end of a massless
> /// cord.
> ```

FYI Probably a bit pedantic, but "cord" implies non-rigid (possible extension and compression). Would this better as a "rigid rod"? (And the change would percolate down to all the other uses).

---

*[drake/examples/multibody/pendulum/pendulum_plant.h, line 43 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-8jKv4TZ--BASqD-zt:-L-8jKv4TZ--BASqD-zu:b-qvzyy6) ([raw file](https://github.com/robotlocomotion/drake/blob/16110aa9a756b18ab09c9ca2240cd44113235fc4/drake/examples/multibody/pendulum/pendulum_plant.h#L43)):*
BTW This last sentence feels a bit awkward. How about:

> This constructor doesn't register any geometry.

(This form parallels the language used in the next constructors documentation).

---

*[drake/examples/multibody/pendulum/pendulum_plant.h, line 46 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-8jACyAJzKRxCXjXCn:-L-8jACyAJzKRxCXjXCo:bixk5g1) ([raw file](https://github.com/robotlocomotion/drake/blob/16110aa9a756b18ab09c9ca2240cd44113235fc4/drake/examples/multibody/pendulum/pendulum_plant.h#L46)):*
> ```objc
>   PendulumPlant(double mass, double length, double gravity);
> 
>   /// Constructs a model of an idealized pendulum wiht parameters as described
> ```

BTW typo: s/wiht/with

---

*[drake/examples/multibody/pendulum/pendulum_plant.h, line 47 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-8jgBziWdT7M_F6s5B:-L-8jgBziWdT7M_F6s5C:b9sqy3c) ([raw file](https://github.com/robotlocomotion/drake/blob/16110aa9a756b18ab09c9ca2240cd44113235fc4/drake/examples/multibody/pendulum/pendulum_plant.h#L47)):*
> ```objc
> 
>   /// Constructs a model of an idealized pendulum wiht parameters as described
>   /// in this class's documentation.
> ```

BTW This constructor and the previous constructor take quite different tacks in describing the parameters. I'd suggest they both do the same thing.

---

*[drake/examples/multibody/pendulum/pendulum_plant.h, line 72 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-8kS5seL00G6oDStAm:-L-8kS5seL00G6oDStAn:b-2vj36w) ([raw file](https://github.com/robotlocomotion/drake/blob/16110aa9a756b18ab09c9ca2240cd44113235fc4/drake/examples/multibody/pendulum/pendulum_plant.h#L72)):*
> ```objc
>   /// Returns the unique id identifying this plant as a source for a
>   /// GeometrySystem.
>   /// The returned id will be invalid when the plant is not registered with any
> ```

The "invalid" state of the identifier is really to allow for default initialization. It isn't intended to be used as a sentinel value in cases like this. You'd be better off returning an `optional<SourceId>` in that case.

See https://github.com/RobotLocomotion/drake/blob/master/drake/geometry/identifier.h#L58

---

*[drake/examples/multibody/pendulum/pendulum_plant.h, line 77 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-8kziKEIB2yaC6XELX:-L-8kziKEIB2yaC6XELY:b-ik4m7u) ([raw file](https://github.com/robotlocomotion/drake/blob/16110aa9a756b18ab09c9ca2240cd44113235fc4/drake/examples/multibody/pendulum/pendulum_plant.h#L77)):*
> ```objc
> 
>   /// Returns the output port of frame id's used to communicate poses to a
>   /// GeometrySystem. It throws a std::out_of_range exception is this system was
> ```

BTW s/is/if

---

*[drake/examples/multibody/pendulum/pendulum_plant.h, line 81 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-8l1igxw3xkBsuCook:-L-8l1igxw3xkBsuCool:b-kfwylt) ([raw file](https://github.com/robotlocomotion/drake/blob/16110aa9a756b18ab09c9ca2240cd44113235fc4/drake/examples/multibody/pendulum/pendulum_plant.h#L81)):*
> ```objc
>   const systems::OutputPort<T>& get_geometry_id_output_port() const;
> 
>   /// Returns the output port of frames's poses to communicate with a
> ```

BTW typo s/frames's'/frames'

---

*[drake/examples/multibody/pendulum/pendulum_plant.h, line 82 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-8l51f1QOcN7EBoGRs:-L-8l51gW5mauvFJKfA6:b-bx56c6) ([raw file](https://github.com/robotlocomotion/drake/blob/16110aa9a756b18ab09c9ca2240cd44113235fc4/drake/examples/multibody/pendulum/pendulum_plant.h#L82)):*
> ```objc
> 
>   /// Returns the output port of frames's poses to communicate with a
>   /// GeometrySystem. It throws a std::out_of_range exception is this system was
> ```

BTW typo s/is/if

---

*[drake/examples/multibody/pendulum/pendulum_plant.h, line 87 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-8lCHggK8PSMKvvYVM:-L-8lCHggK8PSMKvvYVN:b65asdu) ([raw file](https://github.com/robotlocomotion/drake/blob/16110aa9a756b18ab09c9ca2240cd44113235fc4/drake/examples/multibody/pendulum/pendulum_plant.h#L87)):*
> ```objc
> 
>   /// Sets the state for this system in `context` to be that of `this` pendulum
>   /// at a given `angle`. Mainly used to set initial conditions.
> ```

FYI Worth being pedantic in providing the units for `angle`? I generally like to.

---

*[drake/examples/multibody/pendulum/pendulum_plant.h, line 92 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-8lSNAhjyXAOZOsInC:-L-8lSNAhjyXAOZOsInD:b-rh6j1n) ([raw file](https://github.com/robotlocomotion/drake/blob/16110aa9a756b18ab09c9ca2240cd44113235fc4/drake/examples/multibody/pendulum/pendulum_plant.h#L92)):*
> ```objc
>   /// Computes the period for the small oscillations of a simple pendulum.
>   /// The solution assumes no damping. Useful to compute a reference time scale.
>   static T SimplePendulumPeriod(const T& length, const T& gravity) {
> ```

FYI Making *this* depend on the scalar seems overkill.

---

*[drake/examples/multibody/pendulum/pendulum_plant.h, line 154 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-8ltI8QLDtATB2NdQD:-L-8ltI8QLDtATB2NdQE:b-bknmch) ([raw file](https://github.com/robotlocomotion/drake/blob/16110aa9a756b18ab09c9ca2240cd44113235fc4/drake/examples/multibody/pendulum/pendulum_plant.h#L154)):*
> ```objc
> 
>   // The physical parameters of the model:
>   double mass_{1.0};
> ```

FYI Units?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7172)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
A bunch of largely cosmetic things. Probably a couple of discussions.

---

Review status: all files reviewed at latest revision, 27 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-:-L-90-TY8C3Zg_jUkQmo:bj9h6ot)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
BTW As I was coming to work this morning, I realized that some of the perceived awkwardness I pointed out in constructors is due to making the scalar-conversion copy constructor as simple as possible. I pointed out one specific aspect in my comments below. This morning, another occurred to me -- the question of instantiating the geometry ports. I realize now that you have a test in the declaration of geometry output ports and blindly invoke it in the base constructor so that the scalar conversion constructor will have output ports (as necessary). As with the source/frame ids, I'd advocate that the instantiation of ports be conditionally invoked in the body of that constructor rather than paying the cost of making three to four other methods seem a bit weird all to make the scalar conversion constructor as simple as possible.

---

Review status: all files reviewed at latest revision, 27 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-:-L-CMGrWNydJ6u_KHbOO:b-zaccm7)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[drake/examples/multibody/pendulum/passive_simulation.cc, line 12 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-8yWYgsCpJukrU8UGl:-L-COZ0uf-Oyu42UmyXn:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/16110aa9a756b18ab09c9ca2240cd44113235fc4/drake/examples/multibody/pendulum/passive_simulation.cc#L12)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

Unused? It's not included in your command-line options.
</blockquote></details>

Done.

---

*[drake/examples/multibody/pendulum/passive_simulation.cc, line 28 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-8yNpjh_BfdgMY3ZUj:-L-CPtNxMe8hWDVeCYtU:b-jvwjop) ([raw file](https://github.com/robotlocomotion/drake/blob/16110aa9a756b18ab09c9ca2240cd44113235fc4/drake/examples/multibody/pendulum/passive_simulation.cc#L28)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

FYI "playback speed" seems to be a misnomer (particularly given the flag name).
</blockquote></details>

Agree. A copy paste in a lot (all?) demos in master. I changed it here.

---

*[drake/examples/multibody/pendulum/passive_simulation.cc, line 48 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-8ygm6QQpDwewqP_pO:-L-CRDcJqe07q9-RG4Da:bm4na2n) ([raw file](https://github.com/robotlocomotion/drake/blob/16110aa9a756b18ab09c9ca2240cd44113235fc4/drake/examples/multibody/pendulum/passive_simulation.cc#L48)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW `auto*` would make it more apparent that `geometry_system` is a pointer.

Alternatively,

`auto& geometry_system = *builder.AddSystem<GeometrySYstem<double.>();` 

means future invocations of it being passed would look like: `&geometry_system` (consistent with our whole "return a reference" movement).
</blockquote></details>

I like it. I do not like how the call `*builder.AddSystem()` looks like but I do like everything else better as you suggested it.

---

*[drake/examples/multibody/pendulum/passive_simulation.cc, line 128 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-8zhTIJrCW0Z56o1tH:-L-CRlabI1LpuI3pOilO:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/16110aa9a756b18ab09c9ca2240cd44113235fc4/drake/examples/multibody/pendulum/passive_simulation.cc#L128)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

FYI Might be worth including the actual value of the string you got, particularly as this is case sensitive.
</blockquote></details>

Done.

---

*[drake/examples/multibody/pendulum/pendulum_plant.cc, line 64 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-8okUf_wIo5qwmI-rN:-L-CT4Huays1aRqpAuyc:bh2q9mv) ([raw file](https://github.com/robotlocomotion/drake/blob/16110aa9a756b18ab09c9ca2240cd44113235fc4/drake/examples/multibody/pendulum/pendulum_plant.cc#L64)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

Seems that this happens twice -- once here, and once in the next constructor (which is indirectly invoked by *this* constructor). 
</blockquote></details>

Per DM suggestion. Done.

---

*[drake/examples/multibody/pendulum/pendulum_plant.cc, line 75 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-8o6Quqi5BjvtrNl0o:-L-CTLgzC6Wv6YhqBFM3:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/16110aa9a756b18ab09c9ca2240cd44113235fc4/drake/examples/multibody/pendulum/pendulum_plant.cc#L75)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

FYI This comment seems overkill; it's almost a word-for-word repeat of the function name.
</blockquote></details>

Done.

---

*[drake/examples/multibody/pendulum/pendulum_plant.cc, line 87 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-8pA7pfrjtkMlolMeR:-L-CTOhltCumr9sCIv7j:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/16110aa9a756b18ab09c9ca2240cd44113235fc4/drake/examples/multibody/pendulum/pendulum_plant.cc#L87)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

As this constructor knows nothing about geometry, should the output ports exist if it isn't registering geometry?
</blockquote></details>

Done.

---

*[drake/examples/multibody/pendulum/pendulum_plant.cc, line 92 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-8pQgI03Dfhc5sM6Hi:-L-CTQJV00j6Np-zLtvY:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/16110aa9a756b18ab09c9ca2240cd44113235fc4/drake/examples/multibody/pendulum/pendulum_plant.cc#L92)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW Related to my previous comment; if this is only invoked when a frame is *known* to exist, this test becomes superfluous.
</blockquote></details>

Done.

---

*[drake/examples/multibody/pendulum/pendulum_plant.cc, line 144 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-8q7jbuprkXDi3uKF0:-L-CTUAOU0sMziowNuHn:b1nvmcd) ([raw file](https://github.com/robotlocomotion/drake/blob/16110aa9a756b18ab09c9ca2240cd44113235fc4/drake/examples/multibody/pendulum/pendulum_plant.cc#L144)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

FYI Is this an older relic? `GeometrySystem` in master *should* support `AutoDiffXd`.
</blockquote></details>

It is! fixed now.

---

*[drake/examples/multibody/pendulum/pendulum_plant.cc, line 180 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-8qYJ_v4tz9elsWVF-:-L-CYBI0vbExOvhuLvjq:bgvekpx) ([raw file](https://github.com/robotlocomotion/drake/blob/16110aa9a756b18ab09c9ca2240cd44113235fc4/drake/examples/multibody/pendulum/pendulum_plant.cc#L180)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

FYI You have three constructors. The private constructor exists so you can do this. Alternatively, you could eliminate the private constructor if you move its contents into the three-parameter constructor and then modify this so that it's body is:

```
source_id_ = other.source_id_;
frame_id_ = other.frame_id_;
```

The whole extra constructor and additional level of indirection seems a bit much to avoid these two lines.
</blockquote></details>

I reshuffled the whole construction strategy. PTAL

---

*[drake/examples/multibody/pendulum/pendulum_plant.cc, line 189 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-8v9DAdzRpDPDRgWge:-L-CYVrZjJj7xgmgONYi:b-rfog4p) ([raw file](https://github.com/robotlocomotion/drake/blob/16110aa9a756b18ab09c9ca2240cd44113235fc4/drake/examples/multibody/pendulum/pendulum_plant.cc#L189)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW The phrase: "where the rotation axis is" met unhappiness over here. How about:

> ...at the world's origin Wo, through which the rotation axis passes.
</blockquote></details>

much better, thanks.

---

*[drake/examples/multibody/pendulum/pendulum_plant.h, line 18 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-8gsKqJZTKEeY-IqRI:-L-CYzL1Qj0HtS1I_JzK:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/16110aa9a756b18ab09c9ca2240cd44113235fc4/drake/examples/multibody/pendulum/pendulum_plant.h#L18)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

FYI Probably a bit pedantic, but "cord" implies non-rigid (possible extension and compression). Would this better as a "rigid rod"? (And the change would percolate down to all the other uses).
</blockquote></details>

Done.

---

*[drake/examples/multibody/pendulum/pendulum_plant.h, line 72 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-8kS5seL00G6oDStAm:-L-CbL-6R3A-bKDsKcLV:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/16110aa9a756b18ab09c9ca2240cd44113235fc4/drake/examples/multibody/pendulum/pendulum_plant.h#L72)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

The "invalid" state of the identifier is really to allow for default initialization. It isn't intended to be used as a sentinel value in cases like this. You'd be better off returning an `optional<SourceId>` in that case.

See https://github.com/RobotLocomotion/drake/blob/master/drake/geometry/identifier.h#L58
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7172)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Thank you for the recommendations on structuring those constructors @seancurtis-tri, I like it much better now. Back to you.

---

Review status: 0 of 4 files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-:-L-CeBxcM-n4cyueXLeX:bnoiu4f)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
+@sherm1 for platform review please.

---

Review status: 0 of 4 files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-:-L-CgykakrRTdfLTDiO9:b6vi582)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Platform <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with a few localized comments.

---

Reviewed 4 of 8 files at r2, 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 20 unresolved discussions.

---

*[drake/examples/multibody/pendulum/passive_simulation.cc, line 32 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-DmidJwN4tuPf9w0aM:-L-DmidJwN4tuPf9w0aN:b-6mznb6) ([raw file](https://github.com/robotlocomotion/drake/blob/896f6ec5c23aa019f99e1ad86fbca53b05a3a055/drake/examples/multibody/pendulum/passive_simulation.cc#L32)):*
> ```cpp
> DEFINE_string(integration_scheme, "runge_kutta3",
>               "Integration scheme to be used. Available options are:"
>               "'runge_kutta3','implicit_euler','semi_explicit_euler'");
> ```

BTW I tried `./passive_simulation --help` and got the message:
```
passive_simulation: Warning: SetUsageMessage() never called
```
That was followed by the usage message anyway but maybe you should call that function?

---

*[drake/examples/multibody/pendulum/passive_simulation.cc, line 83 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-DndTYpYyHOULGeRPL:-L-DndTYpYyHOULGeRPM:b-u8vrb9) ([raw file](https://github.com/robotlocomotion/drake/blob/896f6ec5c23aa019f99e1ad86fbca53b05a3a055/drake/examples/multibody/pendulum/passive_simulation.cc#L83)):*
> ```cpp
>   PoseBundleToDrawMessage* converter =
>       builder.template AddSystem<PoseBundleToDrawMessage>();
>   LcmPublisherSystem* publisher =
> ```

BTW above you used references rather than pointers to hold the result of AddSystem(). Would be better to do that consistently here also.
(AddSystem() should be fixed to return a reference -- I missed that one .)

---

*[drake/examples/multibody/pendulum/passive_simulation.cc, line 90 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-DohIXlCqbYaazPxaz:-L-DohIXlCqbYaazPxb-:b55mf31) ([raw file](https://github.com/robotlocomotion/drake/blob/896f6ec5c23aa019f99e1ad86fbca53b05a3a055/drake/examples/multibody/pendulum/passive_simulation.cc#L90)):*
> ```cpp
> 
>   // Sanity check on the availability of the optional source id before using it.
>   DRAKE_DEMAND(!!pendulum.get_source_id());
> ```

BTW personally I find these double-negatives hard to understand. I  would prefer `DRAKE_DEMAND(pendulum.get_source_id().has_value())` which seems much clearer. No defect, just please consider changing this.

---

*[drake/examples/multibody/pendulum/passive_simulation.cc, line 117 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-DpZAjnABN6aEMjjof:-L-DpZAjnABN6aEMjjog:bh54lnm) ([raw file](https://github.com/robotlocomotion/drake/blob/896f6ec5c23aa019f99e1ad86fbca53b05a3a055/drake/examples/multibody/pendulum/passive_simulation.cc#L117)):*
> ```cpp
>   systems::Simulator<double> simulator(*diagram, std::move(diagram_context));
> 
>   systems::IntegratorBase<double>* integrator;
> ```

BTW must default-initialize or `= nullptr`.

---

*[drake/examples/multibody/pendulum/passive_simulation.cc, line 131 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-DqAkcdaBky2WXSrKp:-L-DqAkcdaBky2WXSrKq:bsvu6i1) ([raw file](https://github.com/robotlocomotion/drake/blob/896f6ec5c23aa019f99e1ad86fbca53b05a3a055/drake/examples/multibody/pendulum/passive_simulation.cc#L131)):*
> ```cpp
>             *diagram, max_time_step, &simulator.get_mutable_context());
>   } else {
>     throw std::logic_error(
> ```

BTW this should be `std::runtime_error`. For a good explanation of the difference see the answer [here](https://stackoverflow.com/questions/2924058/confused-about-stdruntime-error-vs-stdlogic-error).

---

*[drake/examples/multibody/pendulum/pendulum_plant.cc, line 166 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-Dwm3luK8km4d2tCL_:-L-Dwm3luK8km4d2tCLa:b-8gm58j) ([raw file](https://github.com/robotlocomotion/drake/blob/896f6ec5c23aa019f99e1ad86fbca53b05a3a055/drake/examples/multibody/pendulum/pendulum_plant.cc#L166)):*
> ```cpp
>   DRAKE_DEMAND(model_ != nullptr);
> 
>   // Pose of the com of the pendulum's body (in this case a point mass) in the
> ```

BTW this is a "position" rather than a "pose" since it is just a point location (no axes).

---

*[drake/examples/multibody/pendulum/pendulum_plant.cc, line 209 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-DxRj_oS0uzkH8bWk8:-L-DxRj_oS0uzkH8bWk9:b3f9dey) ([raw file](https://github.com/robotlocomotion/drake/blob/896f6ec5c23aa019f99e1ad86fbca53b05a3a055/drake/examples/multibody/pendulum/pendulum_plant.cc#L209)):*
> ```cpp
>   // Pose of the pendulum's point mass in the body's frame.
>   const Vector3<double> p_BoBcm_B = -get_length() * Vector3<double>::UnitZ();
>   const Isometry3<double> X_BG{Translation3<double>(p_BoBcm_B)};
> ```

BTW why "G" here? Oh, maybe "geometry"? Could use a short comment.

---

*[drake/examples/multibody/pendulum/pendulum_plant.cc, line 269 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-DzkbUEXREQ-8PBgAL:-L-DzkbUEXREQ-8PBgAM:b5k3yzc) ([raw file](https://github.com/robotlocomotion/drake/blob/896f6ec5c23aa019f99e1ad86fbca53b05a3a055/drake/examples/multibody/pendulum/pendulum_plant.cc#L269)):*
> ```cpp
>   model_->CalcInverseDynamics(
>       context, pc, vc, vdot, Fapplied_Bo_W_array, tau_applied,
>       &A_WB_array, &Fapplied_Bo_W_array, &C);
> ```

This confused me. I think now you are ignoring the returned reaction forces by supplying a "dummy" return argument. But that return argument is the same as the applied forces input argument. I don't see anything in the CalcInverseDynamics() parameter descriptions that suggests this is allowed -- what if CalcInverseDynamics() wrote on the output before consuming all the input? What if you need to change the algorithm at some point? 

I can think of a number of ways to fix this. The method could take nullptr for return arguments you don't care about. Or, you could allocate a new array here for holding the unwanted reactions as you did for the unwanted accelerations. Or, if you really want to do this, there should be a very clear comment here about what's going on.

(I did eventually find the `@warning` in CalcInverseDynamics() that says something about these being the same variable. Having seen this in use now I think that's a bad idea.)

---

*[drake/examples/multibody/pendulum/pendulum_plant.h, line 37 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-DuQ8oXiWWH6-vaJ0i:-L-DuQ8oXiWWH6-vaJ0j:b-3eqrqb) ([raw file](https://github.com/robotlocomotion/drake/blob/896f6ec5c23aa019f99e1ad86fbca53b05a3a055/drake/examples/multibody/pendulum/pendulum_plant.h#L37)):*
> ```objc
> /// No other values for T are currently supported.
> template<typename T>
> class PendulumPlant final : public systems::LeafSystem<T> {
> ```

BTW does this have to be final? Could mark individual overrides final if you don't support further overrides while leaving the class extendable, so for example someone could make simple extensions like:
```c++
class BigPendulumPlant : public PendulumPlant {
public:
  BigPendulumPlant() : PendulumPlant(100., 10., 9.81) {}
};
```

---

*[drake/examples/multibody/pendulum/pendulum_plant.h, line 68 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-Dsy4CJPLjAihYSMP4:-L-Dsy4DaNOo7Q2TI-5j:b3r8x8f) ([raw file](https://github.com/robotlocomotion/drake/blob/896f6ec5c23aa019f99e1ad86fbca53b05a3a055/drake/examples/multibody/pendulum/pendulum_plant.h#L68)):*
> ```objc
>   /// Returns the value of the acceleration of gravity in the model.
>   double get_gravity() const { return gravity_; }
> ```

BTW consider getting rid of the `get_` in these names. The style guide allows it (just for the simplest const accessors). @seancurtis-tri and I have recently switched to that and both of us really like it. The names here would then be `mass()`, `length()`, and `gravity()`. `set` methods (if any) would still be `set_mass()`, etc.

---

*[drake/examples/multibody/pendulum/pendulum_plant.h, line 96 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-Dtr2Dy2DpziiO803Z:-L-Dtr2Dy2DpziiO803_:btuvqkb) ([raw file](https://github.com/robotlocomotion/drake/blob/896f6ec5c23aa019f99e1ad86fbca53b05a3a055/drake/examples/multibody/pendulum/pendulum_plant.h#L96)):*
> ```objc
>   /// Computes the period for the small oscillations of a simple pendulum.
>   /// The solution assumes no damping. Useful to compute a reference time scale.
>   static T SimplePendulumPeriod(const T& length, const T& gravity) {
> ```

BTW this method needs  verb, like `CalcSimplePendulumPeriod()`.

---

*[drake/examples/multibody/pendulum/pendulum_plant.h, line 150 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-DvGZIqWVI0kvla4AU:-L-DvGZIqWVI0kvla4AV:b-hte9tq) ([raw file](https://github.com/robotlocomotion/drake/blob/896f6ec5c23aa019f99e1ad86fbca53b05a3a055/drake/examples/multibody/pendulum/pendulum_plant.h#L150)):*
> ```objc
>   double mass_{1.0};      // In kilograms.
>   double length_{0.5};    // In meters.
>   double gravity_{9.81};  // In m/s².
> ```

BTW this class doesn't allow default construction so these initializers are never used but reading the code I would have expected them to be. Should either allow default construction or make the initial values here clearly not usable, like zeroes or NaNs. You could just default-initialize like `double mass_{}` which will set them to zero but doesn't make them look like they have good defaults.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-:-L-E0yPLy5PhdZ00dQNC:b-y7g6kg)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
<img class="emoji" title=":LGTM:" alt=":LGTM:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> along with Sherm's comments.

---

Reviewed 4 of 8 files at r2, 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 16 unresolved discussions.

---

*[drake/examples/multibody/pendulum/passive_simulation.cc, line 90 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-DohIXlCqbYaazPxaz:-L-E4Eq8hRG9GalkhkJj:b-vvgt7g) ([raw file](https://github.com/robotlocomotion/drake/blob/896f6ec5c23aa019f99e1ad86fbca53b05a3a055/drake/examples/multibody/pendulum/passive_simulation.cc#L90)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW personally I find these double-negatives hard to understand. I  would prefer `DRAKE_DEMAND(pendulum.get_source_id().has_value())` which seems much clearer. No defect, just please consider changing this.
</blockquote></details>

BTW `optional` converts nicely to a bool. Does `DRAKE_DEMAND(pendulum.get_source_id());` not work directly?

---

*[drake/examples/multibody/pendulum/passive_simulation.cc, line 108 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-E4d_XSLb5WB5vpTCs:-L-E4d_XSLb5WB5vpTCt:b-dm6srf) ([raw file](https://github.com/robotlocomotion/drake/blob/896f6ec5c23aa019f99e1ad86fbca53b05a3a055/drake/examples/multibody/pendulum/passive_simulation.cc#L108)):*
> ```cpp
>   geometry::DispatchLoadMessage(geometry_system);
> 
>   auto diagram = builder.Build();
> ```

BTW Another case of capturing a reference you could swap out for consistency.

---

*[drake/examples/multibody/pendulum/passive_simulation.cc, line 131 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-DqAkcdaBky2WXSrKp:-L-E4vTEXt0tJRE7o_uX:b-i9exlv) ([raw file](https://github.com/robotlocomotion/drake/blob/896f6ec5c23aa019f99e1ad86fbca53b05a3a055/drake/examples/multibody/pendulum/passive_simulation.cc#L131)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW this should be `std::runtime_error`. For a good explanation of the difference see the answer [here](https://stackoverflow.com/questions/2924058/confused-about-stdruntime-error-vs-stdlogic-error).
</blockquote></details>

BTW I've never much cared for the definition provided to distinguish the two. But I do agree, I like `runtime_error` here more.

---

*[drake/examples/multibody/pendulum/pendulum_plant.h, line 68 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-Dsy4CJPLjAihYSMP4:-L-E2ModKey9IQzAw2iv:bgr5b79) ([raw file](https://github.com/robotlocomotion/drake/blob/896f6ec5c23aa019f99e1ad86fbca53b05a3a055/drake/examples/multibody/pendulum/pendulum_plant.h#L68)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW consider getting rid of the `get_` in these names. The style guide allows it (just for the simplest const accessors). @seancurtis-tri and I have recently switched to that and both of us really like it. The names here would then be `mass()`, `length()`, and `gravity()`. `set` methods (if any) would still be `set_mass()`, etc.
</blockquote></details>

BTW Amen

---

*[drake/examples/multibody/pendulum/pendulum_plant.h, line 73 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-E2uvl16nYznDiAgPf:-L-E2uvl16nYznDiAgPg:b-cekllt) ([raw file](https://github.com/robotlocomotion/drake/blob/896f6ec5c23aa019f99e1ad86fbca53b05a3a055/drake/examples/multibody/pendulum/pendulum_plant.h#L73)):*
> ```objc
>   /// Returns `nullopt` if `this` plant did not register any geometry.
>   optional<geometry::SourceId> get_source_id() const {
>     if (source_id_.is_valid()) {
> ```

BTW This becomes a might simpler if you simply do: 

`optional<geometry::SourceId> source_id_;`

Of course, you'd have to handle it elsewhere in usage. But then you'd never be calling `is_valid()` on an id, you'd simply do:

```C++
if (source_id_) {
  foo(*source_id);
}
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-:-L-E53M48Ww8l6zeWoV3:bdzh9ah)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 1 of 4 files reviewed at latest revision, 8 unresolved discussions.

---

*[drake/examples/multibody/pendulum/passive_simulation.cc, line 32 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-DmidJwN4tuPf9w0aM:-L-HVNY4o-Oa-uzYZFc6:bgb1q6j) ([raw file](https://github.com/robotlocomotion/drake/blob/896f6ec5c23aa019f99e1ad86fbca53b05a3a055/drake/examples/multibody/pendulum/passive_simulation.cc#L32)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW I tried `./passive_simulation --help` and got the message:
```
passive_simulation: Warning: SetUsageMessage() never called
```
That was followed by the usage message anyway but maybe you should call that function?
</blockquote></details>

BTW. Thank you, I had no idea about this call. It looks like a lot of other examples in Drake forget to call it also. I'll push a separate fix PR for those.

---

*[drake/examples/multibody/pendulum/passive_simulation.cc, line 90 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-DohIXlCqbYaazPxaz:-L-HWog15mBOmS8YeF2O:b-lzfcg8) ([raw file](https://github.com/robotlocomotion/drake/blob/896f6ec5c23aa019f99e1ad86fbca53b05a3a055/drake/examples/multibody/pendulum/passive_simulation.cc#L90)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW `optional` converts nicely to a bool. Does `DRAKE_DEMAND(pendulum.get_source_id());` not work directly?
</blockquote></details>

BTW:
- It is not bool convertible within DRAKE_DEMAND (yes within an `if()` statement) @jwnimmer-tri? would you know why?.
- There is no such a thing like `has_value()` (at least in our version of optional).

---

*[drake/examples/multibody/pendulum/passive_simulation.cc, line 108 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-E4d_XSLb5WB5vpTCs:-L-HZEpGWCUtP308AAQr:b-kap8io) ([raw file](https://github.com/robotlocomotion/drake/blob/896f6ec5c23aa019f99e1ad86fbca53b05a3a055/drake/examples/multibody/pendulum/passive_simulation.cc#L108)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW Another case of capturing a reference you could swap out for consistency.
</blockquote></details>

BTW. `Build()` returns a `std::unique_ptr` to the new Diagram. Interestingly, you can still do `const Diagram<double>& diagram = builder.Build()` and your code would still compile. It would just abort with not even a proper trace of where the problems happens (even in gdb). It took me a few minutes to figure out what was going on. What on earth does the compiler even allow you to do that?!

---

*[drake/examples/multibody/pendulum/pendulum_plant.cc, line 269 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-DzkbUEXREQ-8PBgAL:-L-Ha0_1NZ0GtyLc74tV:bz0q1tv) ([raw file](https://github.com/robotlocomotion/drake/blob/896f6ec5c23aa019f99e1ad86fbca53b05a3a055/drake/examples/multibody/pendulum/pendulum_plant.cc#L269)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

This confused me. I think now you are ignoring the returned reaction forces by supplying a "dummy" return argument. But that return argument is the same as the applied forces input argument. I don't see anything in the CalcInverseDynamics() parameter descriptions that suggests this is allowed -- what if CalcInverseDynamics() wrote on the output before consuming all the input? What if you need to change the algorithm at some point? 

I can think of a number of ways to fix this. The method could take nullptr for return arguments you don't care about. Or, you could allocate a new array here for holding the unwanted reactions as you did for the unwanted accelerations. Or, if you really want to do this, there should be a very clear comment here about what's going on.

(I did eventually find the `@warning` in CalcInverseDynamics() that says something about these being the same variable. Having seen this in use now I think that's a bad idea.)

</blockquote></details>

Notice the `CalcInverseDynamics()` in master takes input cache entries in its arguments. Therefore, it is not the final user-facing API. That method is right now intended to be fast and to reduce memory foot-print. We can always wrap it with additional sugar if we so want it. I don't think that should be part of this PR though. 
I added explanatory comment to help understand what is going on.

---

*[drake/examples/multibody/pendulum/pendulum_plant.h, line 72 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-8kS5seL00G6oDStAm:-L-HbzdNhdXaI9wuYcoM:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/16110aa9a756b18ab09c9ca2240cd44113235fc4/drake/examples/multibody/pendulum/pendulum_plant.h#L72)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Done.

---

*[drake/examples/multibody/pendulum/pendulum_plant.h, line 37 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-DuQ8oXiWWH6-vaJ0i:-L-HbBjWwVOsEQ7A6q8l:bkch8w6) ([raw file](https://github.com/robotlocomotion/drake/blob/896f6ec5c23aa019f99e1ad86fbca53b05a3a055/drake/examples/multibody/pendulum/pendulum_plant.h#L37)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW does this have to be final? Could mark individual overrides final if you don't support further overrides while leaving the class extendable, so for example someone could make simple extensions like:
```c++
class BigPendulumPlant : public PendulumPlant {
public:
  BigPendulumPlant() : PendulumPlant(100., 10., 9.81) {}
};
```
</blockquote></details>

BTW. Maybe. For this example I had decided to mark it final since it's a local example. Maybe in some future if we need it. Mostly here I am interested on experimenting with MBT/system/GeometrySystem.

---

*[drake/examples/multibody/pendulum/pendulum_plant.h, line 73 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-E2uvl16nYznDiAgPf:-L-HdAz5JH7GE0Rej-ik:ba2rjv0) ([raw file](https://github.com/robotlocomotion/drake/blob/896f6ec5c23aa019f99e1ad86fbca53b05a3a055/drake/examples/multibody/pendulum/pendulum_plant.h#L73)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW This becomes a might simpler if you simply do: 

`optional<geometry::SourceId> source_id_;`

Of course, you'd have to handle it elsewhere in usage. But then you'd never be calling `is_valid()` on an id, you'd simply do:

```C++
if (source_id_) {
  foo(*source_id);
}
```
</blockquote></details>

I like it better. Done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7172)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
I addressed the reviews. Back to you @seancurtis-tri and @sherm1.

---

Review status: 1 of 4 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-:-L-Hk9U39kLv7OqaW83-:bxk25zs)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Reviewed 3 of 8 files at r2, 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/examples/multibody/pendulum/passive_simulation.cc, line 90 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-DohIXlCqbYaazPxaz:-L-HsQxf_EY5PGQ8o6Ed:b-c44goj) ([raw file](https://github.com/robotlocomotion/drake/blob/896f6ec5c23aa019f99e1ad86fbca53b05a3a055/drake/examples/multibody/pendulum/passive_simulation.cc#L90)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW:
- It is not bool convertible within DRAKE_DEMAND (yes within an `if()` statement) @jwnimmer-tri? would you know why?.
- There is no such a thing like `has_value()` (at least in our version of optional).
</blockquote></details>

BTW The one last option, which may be more easily read:

`DRAKE_DEMAND(pendulum.get_source_id() != nullopt);`

With the same propagating down through the other `DRAKE_DEMAND` instances.

---

*[drake/examples/multibody/pendulum/passive_simulation.cc, line 108 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-E4d_XSLb5WB5vpTCs:-L-HscvJPZaKT1hg2QAY:b-21uhks) ([raw file](https://github.com/robotlocomotion/drake/blob/896f6ec5c23aa019f99e1ad86fbca53b05a3a055/drake/examples/multibody/pendulum/passive_simulation.cc#L108)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW. `Build()` returns a `std::unique_ptr` to the new Diagram. Interestingly, you can still do `const Diagram<double>& diagram = builder.Build()` and your code would still compile. It would just abort with not even a proper trace of where the problems happens (even in gdb). It took me a few minutes to figure out what was going on. What on earth does the compiler even allow you to do that?!
</blockquote></details>

Excellent point. I'd forgotten that it was a `unique_ptr`. The compiler can't stop you doing that -- if it could, we'd never have memory leaks. 

---

*[drake/examples/multibody/pendulum/pendulum_plant.cc, line 204 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-Ht8C8An3h8brZTQrg:-L-Ht8C8An3h8brZTQrh:b-hh4jyu) ([raw file](https://github.com/robotlocomotion/drake/blob/1456c14bc2781c9da9c05060b28a70022dcc7759/drake/examples/multibody/pendulum/pendulum_plant.cc#L204)):*
> ```cpp
>   // frame W.
>   frame_id_ = geometry_system->RegisterFrame(
>       source_id_.value(),
> ```

FYI Alternatively: `*source_id_` for brevity. But if you prefer `source_id_.value()` for clarity, go with it. (And in subsequent instances, of course.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7172)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/examples/multibody/pendulum/pendulum_plant.cc, line 204 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-L-Ht8C8An3h8brZTQrg:-L-Hz5jEKABe6W_9iz83:b-7uxwqb) ([raw file](https://github.com/robotlocomotion/drake/blob/1456c14bc2781c9da9c05060b28a70022dcc7759/drake/examples/multibody/pendulum/pendulum_plant.cc#L204)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

FYI Alternatively: `*source_id_` for brevity. But if you prefer `source_id_.value()` for clarity, go with it. (And in subsequent instances, of course.)
</blockquote></details>

I like `.value()` better because the `*` without context would make you believe `source_id_` could be a pointer.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7172)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

---

Reviewed 3 of 8 files at r2, 2 of 3 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7172#-:-L-I-Sm0vGmyuviktCEO:b-uwkowd)*
<!-- Sent from Reviewable.io -->
